frame/preimage/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+	// The preimage we are storing.
+	pub preimage: BoundedVec,
+	// A reference counter for who depends on this resource.
+	pub ref_count: RefCount<AccountId, Balance>,
+}
+
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Currency type for this pallet.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// An origin that can request a preimage be placed on-chain without a deposit or fee, or
+		/// manage existing preimages.
+		type ManagerOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Max size allowed for a preimage.
+		type MaxSize: Get<u32>;
+
+		/// The base deposit for placing a preimage on chain.
+		type BaseDeposit: Get<BalanceOf<Self>>;
+
+		/// The per-byte deposit for placing a preimage on chain.
+		type ByteDeposit: Get<BalanceOf<Self>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::generate_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A preimage has been noted.
+		Noted { hash: T::Hash },
+		/// A preimage has been requested.
+		Requested { hash: T::Hash },
+		/// A preimage has ben cleared.
+		Cleared { hash: T::Hash },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Preimage is too large to store on-chain.
+		TooLarge,
+		/// Preimage has already been noted on-chain.
+		AlreadyNoted,
+		/// The user is not authorized to perform this action.
+		NotAuthorized,
+	}
+
+	/// The preimages stored by this pallet.
+	// TODO: Maybe store preimage metadata in its own storage.
+	#[pallet::storage]
+	pub(super) type Preimages<T: Config> = StorageMap<
+		_,
+		Identity, // TODO: Double Check
+		T::Hash,
+		Preimage<BoundedVec<u8, T::MaxSize>, T::AccountId, BalanceOf<T>>,
+	>;
+
+	/// Any outstanding preimage requests.
+	#[pallet::storage]
+	pub(super) type Requests<T: Config> = StorageMap<
+		_,
+		Identity, // TODO: Double Check
+		T::Hash,
+		u32,
+		ValueQuery,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a preimage on-chain.
+		///
+		/// If the preimage was previously requested, no fees or deposits are taken for providing
+		/// the preimage. Otherwise, a deposit is taken proportional to the size of the preimage.
+		#[pallet::weight(0)]
+		pub fn note_preimage(origin: OriginFor<T>, bytes: Vec<u8>) -> DispatchResultWithPostInfo {
+			// We accept a signed origin which will pay a deposit, or a root origin where a deposit
+			// is not taken.
+			let maybe_sender = Self::ensure_signed_or_manager(origin)?;
+			let bounded_vec =
+				BoundedVec::<u8, T::MaxSize>::try_from(bytes).map_err(|()| Error::<T>::TooLarge)?;
```

Comment:
```suggestion
				BoundedVec::<u8, T::MaxSize>::try_from(bytes).map_err(|_| Error::<T>::TooLarge)?;
```

Reply:
I prefer to be explicit over the type we are dismissing.